lib/views/video_view/offline_video_player/offline_video_player.dart

```diff
@@ -0,0 +1,161 @@
+          ? const Center(child: CircularProgressIndicator())
+          : _buildVideoLayout(),
+    );
+  }
+
+// Initialize the video player.
+  Future<void> _initVideoPlayer() async {
+    await _initializeAndSeekPlayer();
+  }
+
+// Initialize the controller manager.
+  void _initializeControllerManager() {
+    _controllerManager = OfflineVideoPlayerControllerManager(localPath: widget.localPath);
+  }
+
+// Initialize the video player and seek to the last progress.
+  Future<void> _initializeAndSeekPlayer() async {
+    if (!mounted) return;
+    try {
+      await _controllerManager.initializePlayer();
+      _setLoadingState(false);
+      await _seekToLastProgress();
+    } catch (error) {
+      _handleError(error, 'Failed to initialize video player');
+    }
+  }
+
+// Seek to the last progress.
+  Future<void> _seekToLastProgress() async {
+    final prefs = await SharedPreferences.getInstance();
+    final progress = prefs.getDouble('progress_${widget.localPath}') ?? 0.0;
+    final position = Duration(
+      seconds: (progress * _controllerManager.videoPlayerController.value.duration.inSeconds).round(),
+    );
+    await _controllerManager.videoPlayerController.seekTo(position);
+  }
+
+
+// Handle errors and update the UI accordingly.
+  void _handleError(Object error, String errorMessage) {
+    if (mounted) {
+      ref
+          .read(videoViewModelProvider)
+          .copyWith(error: AppError(errorMessage, error));
+      _setLoadingState(false);
+    }
+  }
+
+  void _setupProgressListener() {
+    _progressTimer =
+        Timer.periodic(const Duration(seconds: 5), _progressUpdateCallback);
+  }
+
+  void _progressUpdateCallback(Timer timer) async {
+    if (!_isPlayerInitialized()) return;
+    if (_controllerManager.videoPlayerController.value.isPlaying) {
+      final position = _getCurrentPosition();
+      final progress = _calculateProgress(position);
+      await _updateProgress(progress);
+      if (_shouldMarkAsWatched(progress)) {
+        await _markStreamAsWatched();
+      }
+    }
+  }
+
+  bool _isPlayerInitialized() {
+    return _controllerManager.videoPlayerController.value.isInitialized;
+  }
+
+  Duration _getCurrentPosition() {
+    return _controllerManager.videoPlayerController.value.position;
+  }
+
+  double _calculateProgress(Duration position) {
+    final duration = _controllerManager.videoPlayerController.value.duration;
+    return position.inSeconds / duration.inSeconds;
+  }
+
+  Future<void> _updateProgress(double progress) async {
+    final prefs = await SharedPreferences.getInstance();
+    await prefs.setDouble('progress_${widget.localPath}', progress);
+  }
+
+  bool _shouldMarkAsWatched(double progress) {
+    const watchedThreshold = 0.9; // 80%
+    return progress >= watchedThreshold;
+  }
+
+  Future<void> _markStreamAsWatched() async {
+    final prefs = await SharedPreferences.getInstance();
+    await prefs.setBool('watched_${widget.localPath}', true);
+  }
```

Comment:
What do we need this code for? (isn't mark as watched only needed for live streams and VoDs?)